### DISCUSSION TITLE
mulpitle homepage: The request matched multiple endpoints. Matches:

### DISCUSSION THREAD

**Discussion by scil** [OrchardCMS/OrchardCore#11662] [Q&A]
Is there any way to  disable or override `OrchardCore.TheTheme.Controllers.HomeController.Index` ?

```
An unhandled exception occurred while processing the request.
AmbiguousMatchException: The request matched multiple endpoints. Matches:

Me.Application.Reading.Controllers.HomeController.Index (Tk.Application.Reading)
OrchardCore.TheTheme.Controllers.HomeController.Index (TheTheme)
Microsoft.AspNetCore.Routing.Matching.DefaultEndpointSelector.ReportAmbiguity(CandidateState[] candidateState)
```

Update:  same error when visiting '/Admin'

```
An unhandled exception occurred while processing the request.
AmbiguousMatchException: The request matched multiple endpoints. Matches:

OrchardCore.Apis.GraphQL.Controllers.AdminController.Index (OrchardCore.Apis.GraphQL)
OrchardCore.Cors.Controllers.AdminController.Index (OrchardCore.Cors)
OrchardCore.Email.Controllers.AdminController.Index (OrchardCore.Email)
OrchardCore.Roles.Controllers.AdminController.Index (OrchardCore.Roles)
OrchardCore.Lucene.Controllers.AdminController.Index (OrchardCore.Lucene)
OrchardCore.AuditTrail.Controllers.AdminController.Index (OrchardCore.AuditTrail)
OrchardCore.Shortcodes.Controllers.AdminController.Index (OrchardCore.Shortcodes)
OrchardCore.ContentTypes.Controllers.AdminController.Index (OrchardCore.ContentTypes)
OrchardCore.Layers.Controllers.AdminController.Index (OrchardCore.Layers)
OrchardCore.Admin.Controllers.AdminController.Index (OrchardCore.Admin)
OrchardCore.Tenants.Controllers.AdminController.Index (OrchardCore.Tenants)
OrchardCore.Users.Controllers.AdminController.Index (OrchardCore.Users)
OrchardCore.Placements.Controllers.AdminController.Index (OrchardCore.Placements)
OrchardCore.Media.Controllers.AdminController.Index (OrchardCore.Media)
OrchardCore.Themes.Controllers.AdminController.Index (OrchardCore.Themes)
OrchardCore.Settings.Controllers.AdminController.Index (OrchardCore.Settings)
OrchardCore.Queries.Controllers.AdminController.Index (OrchardCore.Queries)
OrchardCore.Recipes.Controllers.AdminController.Index (OrchardCore.Recipes)
Microsoft.AspNetCore.Routing.Matching.DefaultEndpointSelector.ReportAmbiguity(CandidateState[] candidateState)
```

**Comment by jtkech** [comment]
From  a theme, a module or from the app?

Show us how you are configuring your route, and maybe your app startup.

E.g. the simplest way from a module

    using Microsoft.AspNetCore.Mvc;

    namespace OrchardCore.HomeRoute
    {
        [Route("~/")]
        public class HomeController : Controller
        {
            public IActionResult Index()
            {
                return View();
            }
        }
    }

**Reply by scil** [reply]
in my Application , I create a home page route
```
            app.UseEndpoints(endpoints => {
                // endpoints.MapControllers();
                endpoints.MapControllerRoute(
                    name: "default",
                    // pattern: "{route=indexhome}");
                    pattern: "{controller=Home}/{action=Index}/{id?}");
            });

```

It cause this error:
```
AmbiguousMatchException: The request matched multiple endpoints. Matches:

Tk.Application.Reading.Controllers.HomeController.Index (Tk.Application.Reading)
OrchardCore.TheTheme.Controllers.HomeController.Index (TheTheme)
OrchardCore.TheTheme.Controllers.HomeController.Index (TheTheme)
Tk.Application.Reading.Controllers.HomeController.Index (Tk.Application.Reading)

Microsoft.AspNetCore.Routing.Matching.DefaultEndpointSelector.ReportAmbiguity(CandidateState[] candidateState)
```


If I rename my controller from Home to AppHome, `pattern: "{controller=AppHome}/{action=Index}/{id?}")`, then  
only two ambiguous matchs:
```
AmbiguousMatchException: The request matched multiple endpoints. Matches:

Tk.Application.Reading.Controllers.HomeController.Index (Tk.Application.Reading)
OrchardCore.TheTheme.Controllers.HomeController.Index (TheTheme)

Microsoft.AspNetCore.Routing.Matching.DefaultEndpointSelector.ReportAmbiguity(CandidateState[] candidateState)
```

**Reply by jtkech** [reply]
In fact in Orchard most of the things are running in the context of a tenant container (we have at least the "Default" tenant) which is composed of modules/features that define area (area name = module Id = module assembly name). In that context you don't need to do `.UseRouting()`, `.UseEndpoints()` and so on, it is done for you, and we are used to define routes in the module/feature startups by using `MapAreaControllerRoute()` to differentiate per area (module assembly name).

In a module we can also define routes by controller/action attributes as in my above example, the controller will be recognized only if the module is enabled, the controller can also be tied to a module feature by using the `Feature` attribute, in that case the related feature should be enabled, and so on.

Yes we can do what we want at the app level but in that case it will bypass what we do at the tenant level, and in your case, because it doesn't differentiate with an area, it is in conflicts with other routes.

Finally, the application itself is a module whose area name is the application assembly name, and we have helpers that act on our `OrchardCoreBuilder` to register things from the application but at the tenant level.

From memory, something like this

    .AddOrchardCms()
    .ConfigureServices(tenantServices => tenantServices
    {
        tenantServices.Add...
    })
    .Configure((builder, routes...) =>
    {
        routes.MapAreaControllerRoute(...
            areaName: "YourAppAssemblyName",
    });

The simplest way to start being my above example using a route attribute, the controller/view being defined in a module, a theme, or in the application (without having to use our helpers).

**Reply by scil** [reply]
Thank you very much.

I just found that there is a recipe file defining `HomeRoute` [HomeRoute](https://docs.orchardcore.net/en/latest/docs/reference/modules/Recipes/#settings-step)

many ways to Rome, it's a little rough for a team to collab.
